tests/e2e-evm/test/abi_basic.test.ts

```diff
@@ -0,0 +1,162 @@
+          expect(code).to.not.equal(0);
+        });
+
+        it("has nonce default of 1", async function () {
+          const nonce = await publicClient.getTransactionCount({ address: mockAddress });
+          expect(nonce).to.equal(1);
+        });
+
+        it("has a starting balance of 0", async function () {
+          const balance = await publicClient.getBalance({ address: mockAddress });
+          expect(balance).to.equal(0n);
+        });
+      });
+
+      for (const funcDesc of abi) {
+        if (funcDesc.type !== "function") {
+          continue;
+        }
+
+        const funcSelector = toFunctionSelector(toFunctionSignature(funcDesc));
+
+        describe(`ABI function ${funcDesc.name} ${funcDesc.stateMutability}`, function () {
+          const isPayable = funcDesc.stateMutability === "payable";
+
+          it("can be called", async function () {
+            const txData = { to: mockAddress, data: funcSelector, gas: 25000n };
+
+            await expect(publicClient.call(txData)).to.be.fulfilled;
+
+            const txHash = await walletClient.sendTransaction(txData);
+            const txReceipt = await publicClient.waitForTransactionReceipt({ hash: txHash });
+            expect(txReceipt.status).to.equal("success");
+          });
+
+          it(`can ${isPayable ? "" : "not "}be called with value`, async function () {
+            const txData = { to: mockAddress, data: funcSelector, gas: 25000n, value: 1n };
+
+            const txHash = await walletClient.sendTransaction(txData);
+            const txReceipt = await publicClient.waitForTransactionReceipt({ hash: txHash });
+            expect(txReceipt.status).to.equal(isPayable ? "success" : "reverted");
+          });
+
+          it("can be called with extra data", async function () {
+            const data = concat([funcSelector, "0x01"]);
+            const txData = { to: mockAddress, data: data, gas: 25000n };
+
+            const txHash = await walletClient.sendTransaction(txData);
+            const txReceipt = await publicClient.waitForTransactionReceipt({ hash: txHash });
+            expect(txReceipt.status).to.equal("success");
+          });
+        });
+      }
+
+      //
+      // Test ABI special functions -- receive & fallback
+      //
+      // Receive functions are always payable and can not receive data
+      // Fallback functions can be payable or non-payable and can receive data in both cases
+      const testName = `ABI special functions: ${receiveFunction ? "" : "no "}receive and ${fallbackFunction ? fallbackFunction.stateMutability : "no"} fallback`;
+      describe(testName, function () {
+        if (!receiveFunction && (!fallbackFunction || fallbackFunction.stateMutability !== "payable")) {
+          it("can not receive plain transfers", async function () {
+            const txData = { to: mockAddress, gas: 25000n, value: 1n };
+
+            const txHash = await walletClient.sendTransaction(txData);
```

Comment:
I've been meaning to refactor this duplication out and create much more concise tests but I've deferred for now since I haven't landed on an API I like.